test_vector_handlers/src/awses_test_vectors/manifests/mpl_keyring.py

```diff
@@ -0,0 +1,182 @@
+    :param str provider_id: Master key provider ID
+    :param str encryption_algorithm: Wrapping key encryption algorithm (required for raw master keys)
+    :param str padding_algorithm: Wrapping key padding algorithm (required for raw master keys)
+    :param str padding_hash: Wrapping key padding hash (required for raw master keys)
+    """
+
+    def keyring(self, keys_uri, mode):
+        # type: (KeysManifest) -> IKeyring
+        """Build a keyring using this specification.
+        :param str keys_uri: Path to the keys manifest
+        """
+        keyvectors = KeyVectorsProvider.get_keyvectors(keys_path=keys_uri)
+
+        # Variable to flag whether we changed anything in weird hack #1.
+        # Signals to weird hack #2 whether it should execute.
+        changed_key_name_from_private_to_public = False
+
+        # Construct the input to KeyVectorsConfig
+        input_kwargs = {
+            "type": self.type_name,
+            "key": self.key_name,
+            "provider-id": self.provider_id,
+            "encryption-algorithm": self.encryption_algorithm,
+
+        }
+        if self.padding_algorithm is not None and self.padding_algorithm != "":
+            input_kwargs["padding-algorithm"] = self.padding_algorithm
+        if self.padding_hash is not None:
+            input_kwargs["padding-hash"] = self.padding_hash
+
+        if input_kwargs["type"] == "raw" \
+                and input_kwargs["encryption-algorithm"] == "rsa":
+            # Weird hack #1:
+            # Gets public key for encryption instead of private key.
+            #
+            # If generating decrypt vectors (i.e. encrypting)
+            # and the manifest specified an RSA private key,
+            # change the input to KeyVectors to a public key.
+            # KeyVectors requires a public key to encrypt.
+            # If this is not done, then keyring.OnEncrypt fails with
+            # "A RawRSAKeyring without a public key cannot provide OnEncrypt"
+            if input_kwargs["key"] == "rsa-4096-private" \
+                    and mode in ("decrypt-generate", "encrypt"):
+                changed_key_name_from_private_to_public = True
+                input_kwargs["key"] = "rsa-4096-public"
+            # Specify default padding-hash
+            if "padding-hash" not in input_kwargs:
+                input_kwargs["padding-hash"] = "sha1"
+
+        # stringify the dict
+        input_as_string = json.dumps(input_kwargs)
+        # convert to unicode code point (expected representation)
+        encoded_json = [ord(c) for c in input_as_string]
+
+        output: GetKeyDescriptionOutput = keyvectors.get_key_description(
+            GetKeyDescriptionInput(json=encoded_json)
+        )
+
+        keyring: IKeyring = keyvectors.create_test_vector_keyring(
+            TestVectorKeyringInput(
+                key_description=output.key_description
+            )
+        )
+
+        # Weird hack #2:
```

Comment:
Weird hack #2 is NOT present in Java.
Java test vectors use a modified keys.json file from other manifests: https://github.com/aws/aws-encryption-sdk-java/blob/master/src/test/resources/keys.json#L45
This is changed from `"key-id": "rsa-4096-private"` to `"key-id": "rsa-4096"`; similarly on the public key.
This allows the `edk.keyProviderInfo == keyring.keyName` check to pass, and the keyring "works".
But (afaik) all of these vector suites provide their own `keys.json` that still have `key-ids` that specify `private`/`public`, so Java would not work with those vector suites.

This code should probably be in the TestVectors Dafny. It's only here because I debugged this in Python, and want some feedback before changing the Dafny.

Reply:
I'd describe it as "coordinating" rather than fighting :)
The primary evidence for this is that the test vectors pass with this change

I think the plan of action here SHOULD be some combination of 1) rename key manifests going forward so the key name of a public/private key pair is the same, 2) add a "legacy" check in TestVectors to align the key name of a public/private key pair

Reply:
Is this plan of action going to happen here or is this happening somewhere else?